text/3594-expose-stackrealign-attribute.md

```diff
@@ -0,0 +1,101 @@
+
+In the example above the "contract" is defined by hardware and very hard\impossible to work around. 
+
+The proposed attribute will tell the compiler that the precondition that the stack is aligned might not be true, and that it might need to fix it up in order to execute the function correctly.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+The `[realign_stack]` attribute can be added to a function to tell the compiler to add stack re-alignment to that function if necessary.
+Useful in cases where your code is called from a thread or a binary compiled with another compiler, that uses different alignment and thus lead to a corruption.
+An example of one such setting could be `-mpreferred-stack-boundary=2` in GCC which would set the stack alignment to 4 instead of the default value for the ABI which is 16.
+Other such settings could be present at GCC's "Machine-Dependent Options", which there are many of, and many of them can break ABI compatibility.
+
+```
+#[realign_stack]
+#[no_mangle]
+pub extern "C" fn callback_function() -> i32 {
+    println!("Called from callback!!");
+
+    0
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+The realign_stack attribute is specified as follows:
+```
+#[realign_stack]
+```
+
+When the `realign_stack` attribute is applied to a function, the compiler no longer assumes the stack is properly aligned when the function is called, and so will insert code to forcefully realign the stack as needed for calling other functions, variables requiring alignment, etc.
+This alignment is achieved by adjusting the stack pointer accordingly to the stack alignment specified in the target ABI's data layout.
+Adding this attribute unnecessarily might "waste" space on the stack which could be crucial in real-time systems.
+
+In LLVM the `alignstack` gets an argument that specifies the alignment in bytes(also must be a power of 2). 
+Below is an example of how it would work for an example data-layout:
+`e-m:e-i64:64-f80:128-n8:16:32:64-S128`.
+The `S128` is the part that describes the natural stack alignment in bits.
+So practically, we just need to divide that value by 8, and place it as the argument of `alignstack`.
+So in the `S128` case it would look like this: `alignstack=16`.
+
+```
+define i32 @callback_function() #0 {
+start:
+  ret i32 0
+}
+
+attributes #0 = { alignstack=16 }
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+- Introducing a new attribute adds complexity to the language.
+- Limited use cases: Stack realignment may not be necessary in most Rust codebases, potentially making this feature less relevant for many developers.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+An alternative could be a macro workaround instead of adding the attribute.

```

Comment:
Yeah.
I meant that maybe this RFC could already be implemented purely in rust already with some naked function and inline assembly shenanigans

Reply:
You've said that you're currently blocked on this issue.  Have you explored this macro idea as a possible solution?  @Amanieu's [`naked_function::naked`](https://docs.rs/naked-function/latest/naked_function/attr.naked.html) macro might be a good starting point?

Reply:
I'm not really good enough with macros and such to do that but yeah I think I will maybe try exploring something along that way.

Reply:
I'd suggest that you start by using that `naked` macro just to implement a stack-realigning prelude that then calls or jumps to your actual logic defined separately in a regular function.

Reply:
@eggyal It should be extremely difficult to do it because editing the stack without really informing the compiler about it should cause many issues. For example arguments on stack are used like `[esp + 0xc]`  or something, and aligning the stack with something like `and esp, -16` will cause `[esp + 0xc]` to bring wrong arguments.

Reply:
You'll have to [copy](https://doc.rust-lang.org/std/ptr/fn.copy.html) whatever is on the misaligned stack to the adjusted, aligned location.

Reply:
In c when it does this realigning, the compiler KNOWS realigning happened, and in those functions a frame pointer is always created and data is retrieved relative to that. I don't know how we can force frame pointers for functions in rust. It's the compilers choice whether to create them or not.

Reply:
@eggyal For reference https://godbolt.org/z/cxvYx58ch
That's how it should like with proper stack realignment. 
You can then comment out the stack realignment and see how it just removes the frame pointer alltogether.
Here https://godbolt.org/z/cKsTdP8xq you can see that rust does not always generate frame pointers and the decision is not left to developers alltogether.
It it chooses to create one BUT we also create one (for example via inline assembly) the compiler will assume the wrong frame pointer is loaded which will yield UB.

Reply:
well, it is less efficient but you can make a thunk in assembly that aligns the stack and passes the arguments to the inner function:
```rust
#[naked]
pub unsafe extern "C" fn my_fn(a: u32) -> i32 {
    asm!(
        "push ebp",
        "mov ebp, esp",
        "and esp, -16", // align to 16 bytes
        "push dword ptr [ebp + 4]", // copy argument
        "call {}",
        "mov esp, ebp",
        "pop ebp",
        "ret",
        sym my_fn_impl,
        options(noreturn),
    );
}

extern "C" fn my_fn_impl(a: u32) -> i32 {
    ...
}
```

Reply:
> It it chooses to create one BUT we also create one (for example via inline assembly) the compiler will assume the wrong frame pointer is loaded which will yield UB.

unless you're writing the entire function in assembly with global asm or naked functions, trying to align the stack pointer using asm is *always* UB (since you're exiting the asm block with a different `esp` than you entered it with), just sometimes that UB happens to do what you want (when LLVM decides to create a frame pointer and doesn't assume more things you weren't expecting).

This is like how signed integer overflow is *always* UB (unless using `-fwrapv` or equivalent) but still sometimes does what you want.

Sometimes doing what you want is how *all* UB works.

Reply:
> I will check it out tomorrow. The problem is I want it to be generic in a way that I could put that attribute on any function and that it should just work.

yeah, I'm not saying we shouldn't have something like `#[realign_stack]` since that is far easier to use, is more efficient, and is more likely to be correct.

> However, there are issues like passing a u64 in 32 bits to a function is actually pushing the higher part and the lower part (IE like 2 function arguments) which makes creating a macro like that difficult (in terms of choosing how much to choose to push from the old stack to the inner function).

you can make a proc macro that implements some of the C ABI, so for a `u64` argument, you'd push 2 32-bit dwords. the generated asm is specific to each function signature.

Reply:
for `extern "C" fn(i64) -> i64` you can use:

```
        push    ebp
        mov     ebp, esp
        and     esp, -16
        push    dword ptr [ebp + 12]
        push    dword ptr [ebp + 8]
        call    {}
        mov     esp, ebp
        pop     ebp
        ret
```

Reply:
I will try it out tomorrow and see what my team members think about it.

Reply:
@programmerjake We will probably take the proc macro solution in the mean while, mostly because it allows us to upgrade our rust version to a newer version which will yield smaller binary size and has many other improvements as well.
Obviously though we, are really looking forward to this RFC being pushed so no more hacky solution will be needed.

Reply:
one of the parts that makes that particularly easy on x86's C abi is that everything is passed on the stack, so all you need is to figure out how much stack to copy.

Reply:
I had a bit of a hack around and came up with the following, *completely untested*, macro (depends on [naked-function](https://crates.io/crates/naked-function) and [paste](https://crates.io/crates/paste)); my intention was to create something general enough to be applied to any x86-64 Sys V function (future adaptations could expand to other architectures/ABIs), but it comes with some runtime costs:
* two calls (one to discover the number of arguments on the stack, and the other to delegate to the "actual" function once the stack has been aligned—maybe LLVM will inline the latter, but I think the former relies on the inner call to `end_of_stack_args` not being inlined); and
* one loop (that copies the unaligned stack arguments to the aligned stack, unfortunately 16-bits at a time because in 64-bit mode it can't do a 32-bit read and a 64-bit read would be unaligned).

```rust
#[doc(hidden)]
pub mod reexport {
    pub extern crate naked_function;
    pub extern crate paste;
    pub extern crate std;

    #[macro_export]
    #[cfg(all(not(doc), target_arch = "x86_64"))]
    macro_rules! realign {
        (
            $(#[$attr:meta])*
            $vis:vis unsafe extern "C" fn $function:ident($($arg:ident: $argt:ty),*) $(-> $return:ty)? $body:block
        ) => {
            $crate::reexport::paste::paste! {
                #[inline(always)]
                fn [<__number_of_stack_args_ $function>]() -> usize {
                    #[inline(never)]
                    extern "C" fn end_of_stack_args($(_: $crate::reexport::std::mem::MaybeUninit<$argt>)*) -> usize {
                        unsafe {
                            let end_of_stack_args;
                            $crate::reexport::std::arch::asm!(
                                "lea {}, [rbp + 8]",
                                lateout(reg) end_of_stack_args,
                                options(pure, nomem, preserves_flags, nostack),
                            );                
                            end_of_stack_args
                        }
                    }
                    unsafe {
                        let start_of_stack_args: usize;
                        $crate::reexport::std::arch::asm!(
                            "mov {}, rsp",
                            lateout(reg) start_of_stack_args,
                            options(pure, nomem, preserves_flags, nostack),
                        );                
                        (start_of_stack_args - end_of_stack_args($($crate::reexport::std::mem::MaybeUninit::<$argt>::uninit()),*)) / 8
                    }
                }
    
                #[inline(always)]
                unsafe extern "C" fn [<__actual_ $function>]($($arg: $argt)*) $(-> $return)? $body
    
                #[$crate::reexport::naked_function::naked]
                $(#[$attr])*
                $vis unsafe extern "C" fn $function($($arg: $argt)*) $(-> $return)? {
                    asm! {
                        "enter 0, 0",
                        "and rsp, -16",
    
                        "call {num_of_stack_args}",
                        "test rax, rax",
                        "jz endloop",
                        "shl rax, 2",
    
                        "loop:",
                            "push word ptr [rbp + 8 + rax*2]",
                            "dec rax",
                            "jnz loop",
    
                        "endloop:",
                            "call {actual}",
                            "leave",
                            "ret",
    
                        num_of_stack_args = sym [<__number_of_stack_args_ $function>],
                        actual = sym [<__actual_ $function>],
                    }
                }
            }
        };
    }
}

/// Generates a thunk around the contained function that is robust to being called with a
/// stack underaligned relative to that expected by the x86-64 Sys V ABI.
///
/// # Example
/// ```rust
/// use std::ffi::{c_char, CStr};
/// 
/// realign_stack::realign! {
///     /// This function can be called with an underlined stack.
///     unsafe extern "C" fn robust(name: *const c_char) {
///         assert!(!name.is_null());
///         let name = CStr::from_ptr(name);
///         println!("hello {}", name.to_str().unwrap());
///     }
/// }
/// ```
#[cfg(doc)]
#[macro_export]
macro_rules! realign {
    ($function:item) => {};
}
```

Reply:
yeah could be a bit annoying

Reply:
Seems very interesting.
Why did you choose a `macro_rules` and not `proc_macro_attribute`?
Also seems very nice that `paste` crate

Reply:
I just thought of a waay easier workaround that's practically always correct: just guarantee you have a local variable with alignment greater than required by the ABI (so >= 16 bytes for x86), LLVM will then have to align the stack for that variable, no complex inline asm required: https://rust.godbolt.org/z/WaExsWvGj

Reply:
> Why did you choose a `macro_rules` and not `proc_macro_attribute`?

Mostly for ease of implementation, but also is faster to compile.  If you'd prefer the API of an attribute macro, it'd be pretty easy to make one emit an invocation of the macro above.

Reply:
> I just thought of a waay easier workaround that's practically always correct: just guarantee you have a local variable with alignment greater than required by the ABI

But isn't the problem that the arguments are placed on the misaligned stack *before* such forcibly aligned local variable?  I can't see any way to cope with that on stable Rust today other than via `global_asm` (here via `naked-function` crate).  The problem then becomes how one gets from such asm to a function body that has been written in Rust... and alas, I think that has to involve copying the (on-stack) arguments and making a `call`.

Reply:
> > I just thought of a waay easier workaround that's practically always correct: just guarantee you have a local variable with alignment greater than required by the ABI
> 
> But isn't the problem that the arguments are placed on the misaligned stack _before_ such forcibly aligned local variable? I can't see any way to cope with that on stable Rust today other than via `global_asm` (here via `naked-function` crate). The problem then becomes how one gets from such asm to a function body that has been written in Rust... and alas, I think that has to involve copying the (on-stack) arguments and making a `call`.

make the function called by the trampoline `#[inline(never)]`:
either:
```rust
#[repr(align(32))]
struct Aligned;

pub unsafe extern "C" fn the_fn(a: u32, b: u64, c: &u8) -> u64 {
    let align = Aligned;
    asm!("/* {} */", in(reg) &align);
    the_fn_impl(a, b, c)
}

#[inline(never)]
fn the_fn_impl(a: u32, b: u64, c: &u8) -> u64 {
    todo!()
}
```

or:
```rust
/// workaround unstable attributes on expressions
#[inline(always)]
fn identity<T>(v: T) -> T {
    v
}

#[repr(align(32))]
struct Aligned;

pub unsafe extern "C" fn the_fn(a: u32, b: u64, c: &u8) -> u64 {
    let align = Aligned;
    asm!("/* {} */", in(reg) &align);
    identity(#[inline(never)] || the_fn_impl(a, b, c))()
}

fn the_fn_impl(a: u32, b: u64, c: &u8) -> u64 {
    todo!()
}
```